### PR TITLE
enable upgrades at 0 timestamp in chain configs

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Mirrors: https://github.com/ava-labs/coreth/pull/1133

## How this works

Same as https://github.com/ava-labs/coreth/pull/1133

one difference is that Subnet-EVM validates if the chain config reconfigures any upgrade earlier than the default network upgrades (ones in avalanchego ctx). This is because Subnet-EVM actually allows reconfigurable network upgrades either through override network upgrades or in the genesis configuration. Subnet-EVM was verifying to ensure there is no early activation then avalanche network upgrades. However it was not handling the edge cases (where ctx upgrade is at `UnscheduledTime` or `InitiallyActiveTime` timestamps), so this one properly handles those edge cases to make sure tests are not broken with `0`=`InitiallyActive` and `nil`=`UnscheduledTime` cases.

## How this was tested

UT

## Need to be documented?

No

## Need to update RELEASES.md?

No

### CODE REVIEW

File: core/genesis.go [hunk 1]
```diff
@@ -317,15 +318,31 @@ func (g *Genesis) toBlock(db ethdb.Database, triedb *triedb.Database) *types.Blo
...
 			if g.BaseFee != nil {
 				head.BaseFee = g.BaseFee
 			} else {
 				head.BaseFee = new(big.Int).Set(params.GetExtra(g.Config).FeeConfig.MinBaseFee)
 			}
 		}
+		headerExtra := customtypes.GetHeaderExtra(head)
+
+		// When Etna/Cancun is active, `BlockGasCost` and `ExtDataGasUsed` are decoded to 0 if it's nil.
+		// This is because these fields come before than other optional Cancun fields in RLP order.
```
Comment (alarso16) [inline core/genesis.go:337]: Same typo

File: params/paramstest/forks.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+package paramstest
+
+import (
+	"github.com/ava-labs/avalanchego/upgrade/upgradetest"
+
+	"github.com/ava-labs/subnet-evm/params"
+)
+
+// ForkToChainConfig maps a fork to a chain config
+var ForkToChainConfig = map[upgradetest.Fork]*params.ChainConfig{
```
Comment (JonathanOppenheimer) [inline params/paramstest/forks.go:13]: This is already present in the `var` block of `vm_test.go.` I don't know if should be in a separate file here, or whether we can avoid the duplication altogether. At a gut level it seems silly to create a whole new file for a var block.

File: params/extras/network_upgrades_test.go [hunk 3]
```diff
@@ -284,24 +274,25 @@ func TestVerifyNetworkUpgrades(t *testing.T) {
...
 			err := test.upgrades.verifyNetworkUpgrades(test.avagoUpgrades)
 			if test.valid {
-				require.Nil(t, err)
+				require.NoError(t, err)
 			} else {
-				require.NotNil(t, err)
+				require.Error(t, err)
 			}
 		})
 	}
```
Comment (JonathanOppenheimer) [inline params/extras/network_upgrades_test.go:298]:
```suggestion
	}
}

func TestSetDefaultsTreatsZeroAsUnsetForLiveNetworks(t *testing.T) {
	// Use a real AvalancheGo config with non-unscheduled defaults
	ago := upgradetest.GetConfig(upgradetest.Granite)

	// Provide zeros for all timestamps to simulate "unset" in user config
	n := NetworkUpgrades{
		SubnetEVMTimestamp: utils.NewUint64(0),
		DurangoTimestamp:   utils.NewUint64(0),
		EtnaTimestamp:      utils.NewUint64(0),
		FortunaTimestamp:   utils.NewUint64(0),
		GraniteTimestamp:   utils.NewUint64(0),
	}

	// Apply defaults: zeros should be replaced with AvalancheGo defaults
	n.SetDefaults(ago)

	require.NotNil(t, n.SubnetEVMTimestamp)
	require.NotNil(t, n.DurangoTimestamp)
	require.NotNil(t, n.EtnaTimestamp)
	// Fortuna is optional (may be nil in defaults); we only assert that Granite is set from defaults
	require.NotNil(t, n.GraniteTimestamp)

	defaults := GetNetworkUpgrades(ago)

	require.Equal(t, *defaults.SubnetEVMTimestamp, *n.SubnetEVMTimestamp)
	require.Equal(t, *defaults.DurangoTimestamp, *n.DurangoTimestamp)
	require.Equal(t, *defaults.EtnaTimestamp, *n.EtnaTimestamp)
	// Fortuna default is nil; if defaults has nil keep it nil, otherwise equal
	if defaults.FortunaTimestamp == nil {
		require.Nil(t, n.FortunaTimestamp)
	} else {
		require.NotNil(t, n.FortunaTimestamp)
		require.Equal(t, *defaults.FortunaTimestamp, *n.FortunaTimestamp)
	}
	require.Equal(t, *defaults.GraniteTimestamp, *n.GraniteTimestamp)
```

Comment (JonathanOppenheimer) [reply params/extras/network_upgrades_test.go:298]: Should we test `setDefaults`?  Feel free to modify.

File: core/genesis_test.go [hunk 4]
```diff
@@ -65,7 +65,7 @@ func setupGenesisBlock(db ethdb.Database, triedb *triedb.Database, genesis *Gene
 }
```
Comment (JonathanOppenheimer) [inline core/genesis_test.go:66]:
```suggestion
func TestGenesisRoundTripNormalization_EtnaCancun(t *testing.T) {
	require := require.New(t)

	cfg := params.Copy(params.TestEtnaChainConfig)
	// Ensure Subnet-EVM active at genesis and Cancun active at genesis via params.TestEtnaChainConfig
	// BaseFee should be set if missing; ParentBeaconRoot zero; optional Cancun fields present and zero.

	g := &Genesis{
		Config:     &cfg,
		Timestamp:  0,
		Difficulty: big.NewInt(0),
		GasLimit:   params.GetExtra(&cfg).FeeConfig.GasLimit.Uint64(),
		// Leave ExtraData nil to assert it normalizes to empty slice
		// Leave BaseFee nil to assert it is set from FeeConfig.MinBaseFee
	}

	db := rawdb.NewMemoryDatabase()
	trieDB := triedb.NewDatabase(db, triedb.HashDefaults)
	committed := g.MustCommit(db, trieDB)

	// Read header back and assert normalized fields
	hdr := rawdb.ReadHeader(db, committed.Hash(), committed.NumberU64())
	require.NotNil(hdr)

	// ExtraData normalization: nil -> empty []
	assert.NotNil(t, hdr.Extra)
	assert.Equal(t, 0, len(hdr.Extra))

	// BaseFee must be non-nil when Subnet-EVM is active
	assert.NotNil(t, hdr.BaseFee)
	assert.True(t, hdr.BaseFee.Sign() >= 0)

	// Etna normalization: BlockGasCost non-nil and equals 0
	extra := customtypes.GetHeaderExtra(hdr)
	if params.GetExtra(&cfg).NetworkUpgrades.IsEtna(g.Timestamp) {
		require.NotNil(t, extra.BlockGasCost)
		assert.Equal(t, 0, extra.BlockGasCost.Sign())
	}

	// Cancun normalization: ParentBeaconRoot present and zero; blob fields present and zero
	if hdr.ParentBeaconRoot != nil {
		assert.Equal(t, common.Hash{}, *hdr.ParentBeaconRoot)
	}
	require.NotNil(t, hdr.BlobGasUsed)
	assert.Equal(t, uint64(0), *hdr.BlobGasUsed)
	require.NotNil(t, hdr.ExcessBlobGas)
	assert.Equal(t, uint64(0), *hdr.ExcessBlobGas)
}
```

Comment (JonathanOppenheimer) [reply core/genesis_test.go:66]: should also test the normalization?
